test flake: TestSSH/StartStoppedWorkspace

https://github.com/coder/coder/actions/runs/7287471624/job/19858183477

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: panic recovered for coder ssh: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x7ff7a7e88e4d]

goroutine 52988 [running]:
github.com/coder/coder/v2/cli/clibase.(*Invocation).Run.func1()
	C:/a/coder/coder/cli/clibase/cmd.go:502 +0x105
panic({0x7ff7a95d4e40?, 0x7ff7a929eac0?})
	C:/hostedtoolcache/windows/go/1.21.5/x64/src/runtime/panic.go:914 +0x21f
github.com/coder/coder/v2/cli/cliui.Agent({0x7ff7aa9a5a10?, 0xc00f5f1540?}, {0x7ff7aa9838e0?, 0xc00d235480}, {0xe1, 0x6a, 0x4d, 0x48, 0x60, 0x57, ...}, ...)
	C:/a/coder/coder/cli/cliui/agent.go:203 +0x134d
github.com/coder/coder/v2/cli.(*RootCmd).ssh.func1(0xc009a32a00)
	C:/a/coder/coder/cli/ssh.go:199 +0x1385
github.com/coder/coder/v2/cli.(*RootCmd).Command.func6.1(0xc00be0d110?)
	C:/a/coder/coder/cli/root.go:290 +0xbe
github.com/coder/coder/v2/cli.(*RootCmd).ssh.(*RootCmd).InitClient.(*RootCmd).initClientInternal.func3.1(0xc009a32a00)
	C:/a/coder/coder/cli/root.go:607 +0x670
github.com/coder/coder/v2/cli.(*RootCmd).ssh.RequireNArgs.RequireRangeArgs.func2.1(0x0?)
	C:/a/coder/coder/cli/clibase/cmd.go:614 +0x328
github.com/coder/coder/v2/cli/clibase.(*Invocation).run(0xc009a32a00, 0xc012d77f20)
	C:/a/coder/coder/cli/clibase/cmd.go:428 +0x1022
github.com/coder/coder/v2/cli/clibase.(*Invocation).run(0xc009a32300, 0xc01047df20)
	C:/a/coder/coder/cli/clibase/cmd.go:365 +0xae5
github.com/coder/coder/v2/cli/clibase.(*Invocation).Run(0xc002d68600?)
	C:/a/coder/coder/cli/clibase/cmd.go:518 +0xdd
github.com/coder/coder/v2/cli_test.TestSSH.func2.1()
	C:/a/coder/coder/cli/ssh_test.go:121 +0x2d
github.com/coder/coder/v2/cli_test.tGo.func2()
	C:/a/coder/coder/cli/ssh_test.go:1057 +0x23
created by github.com/coder/coder/v2/cli_test.tGo in goroutine 1513
	C:/a/coder/coder/cli/ssh_test.go:1056 +0xcc
```

--- a/cli/cliui/agent.go
+++ b/cli/cliui/agent.go
@@ -238,12 +238,13 @@ func Agent(ctx context.Context, writer io.Writer, agentID uuid.UUID, opts AgentO
 			sw.Log(time.Now(), codersdk.LogLevelWarn, "Wait for it to reconnect or restart your workspace.")
 			sw.Log(time.Now(), codersdk.LogLevelWarn, troubleshootingMessage(agent, "https://coder.com/docs/v2/latest/templates#agent-connection-issues"))
 
+			disconnectedAt := agent.DisconnectedAt
 			for agent.Status == codersdk.WorkspaceAgentDisconnected {
 				if agent, err = fetch(); err != nil {
 					return xerrors.Errorf("fetch: %w", err)
 				}
 			}
-			sw.Complete(stage, safeDuration(sw, agent.LastConnectedAt, agent.DisconnectedAt))
+			sw.Complete(stage, safeDuration(sw, agent.LastConnectedAt, disconnectedAt))
 		}
 	}
 }

--- a/cli/cliui/agent.go
+++ b/cli/cliui/agent.go
@@ -200,12 +200,12 @@ func Agent(ctx context.Context, writer io.Writer, agentID uuid.UUID, opts AgentO
 
 			switch agent.LifecycleState {
 			case codersdk.WorkspaceAgentLifecycleReady:
-				sw.Complete(stage, safeDuration(agent.ReadyAt, agent.StartedAt))
+				sw.Complete(stage, safeDuration(sw, agent.ReadyAt, agent.StartedAt))
 			case codersdk.WorkspaceAgentLifecycleStartTimeout:
 				sw.Fail(stage, 0)
 				sw.Log(time.Time{}, codersdk.LogLevelWarn, "Warning: A startup script timed out and your workspace may be incomplete.")
 			case codersdk.WorkspaceAgentLifecycleStartError:
-				sw.Fail(stage, safeDuration(agent.ReadyAt, agent.StartedAt))
+				sw.Fail(stage, safeDuration(sw, agent.ReadyAt, agent.StartedAt))
 				// Use zero time (omitted) to separate these from the startup logs.
 				sw.Log(time.Time{}, codersdk.LogLevelWarn, "Warning: A startup script exited with an error and your workspace may be incomplete.")
 				sw.Log(time.Time{}, codersdk.LogLevelWarn, troubleshootingMessage(agent, "https://coder.com/docs/v2/latest/templates#startup-script-exited-with-an-error"))
@@ -221,7 +221,7 @@ func Agent(ctx context.Context, writer io.Writer, agentID uuid.UUID, opts AgentO
 				case agent.LifecycleState.ShuttingDown():
 					// We no longer know if the startup script failed or not,
 					// but we need to tell the user something.
-					sw.Complete(stage, safeDuration(agent.ReadyAt, agent.StartedAt))
+					sw.Complete(stage, safeDuration(sw, agent.ReadyAt, agent.StartedAt))
 					return errAgentShuttingDown
 				}
 			}
@@ -243,7 +243,7 @@ func Agent(ctx context.Context, writer io.Writer, agentID uuid.UUID, opts AgentO
 					return xerrors.Errorf("fetch: %w", err)
 				}
 			}
-			sw.Complete(stage, safeDuration(agent.LastConnectedAt, agent.DisconnectedAt))
+			sw.Complete(stage, safeDuration(sw, agent.LastConnectedAt, agent.DisconnectedAt))
 		}
 	}
 }
@@ -262,8 +262,14 @@ func troubleshootingMessage(agent codersdk.WorkspaceAgent, url string) string {
 // which are not critical, and therefor should not break things
 // when it fails.
 // A panic has been observed in a test.
-func safeDuration(a, b *time.Time) time.Duration {
+func safeDuration(sw *stageWriter, a, b *time.Time) time.Duration {
 	if a == nil || b == nil {
+		if sw != nil {
+			// Ideally the message includes which fields are <nil>, but you can
+			// use the surrounding log lines to figure that out. And passing more
+			// params makes this unwieldy.
+			sw.Log(time.Now(), codersdk.LogLevelWarn, "Warning: Failed to calculate duration from a time being <nil>.")
+		}
 		return 0
 	}
 	return a.Sub(*b)

--- a/cli/cliui/agent.go
+++ b/cli/cliui/agent.go
@@ -200,12 +200,12 @@ func Agent(ctx context.Context, writer io.Writer, agentID uuid.UUID, opts AgentO
 
 			switch agent.LifecycleState {
 			case codersdk.WorkspaceAgentLifecycleReady:
-				sw.Complete(stage, agent.ReadyAt.Sub(*agent.StartedAt))
+				sw.Complete(stage, safeDuration(agent.ReadyAt, agent.StartedAt))
 			case codersdk.WorkspaceAgentLifecycleStartTimeout:
 				sw.Fail(stage, 0)
 				sw.Log(time.Time{}, codersdk.LogLevelWarn, "Warning: A startup script timed out and your workspace may be incomplete.")
 			case codersdk.WorkspaceAgentLifecycleStartError:
-				sw.Fail(stage, agent.ReadyAt.Sub(*agent.StartedAt))
+				sw.Fail(stage, safeDuration(agent.ReadyAt, agent.StartedAt))
 				// Use zero time (omitted) to separate these from the startup logs.
 				sw.Log(time.Time{}, codersdk.LogLevelWarn, "Warning: A startup script exited with an error and your workspace may be incomplete.")
 				sw.Log(time.Time{}, codersdk.LogLevelWarn, troubleshootingMessage(agent, "https://coder.com/docs/v2/latest/templates#startup-script-exited-with-an-error"))
@@ -221,7 +221,7 @@ func Agent(ctx context.Context, writer io.Writer, agentID uuid.UUID, opts AgentO
 				case agent.LifecycleState.ShuttingDown():
 					// We no longer know if the startup script failed or not,
 					// but we need to tell the user something.
-					sw.Complete(stage, agent.ReadyAt.Sub(*agent.StartedAt))
+					sw.Complete(stage, safeDuration(agent.ReadyAt, agent.StartedAt))
 					return errAgentShuttingDown
 				}
 			}
@@ -238,13 +238,12 @@ func Agent(ctx context.Context, writer io.Writer, agentID uuid.UUID, opts AgentO
 			sw.Log(time.Now(), codersdk.LogLevelWarn, "Wait for it to reconnect or restart your workspace.")
 			sw.Log(time.Now(), codersdk.LogLevelWarn, troubleshootingMessage(agent, "https://coder.com/docs/v2/latest/templates#agent-connection-issues"))
 
-			disconnectedAt := *agent.DisconnectedAt
 			for agent.Status == codersdk.WorkspaceAgentDisconnected {
 				if agent, err = fetch(); err != nil {
 					return xerrors.Errorf("fetch: %w", err)
 				}
 			}
-			sw.Complete(stage, agent.LastConnectedAt.Sub(disconnectedAt))
+			sw.Complete(stage, safeDuration(agent.LastConnectedAt, agent.DisconnectedAt))
 		}
 	}
 }
@@ -255,6 +254,19 @@ func troubleshootingMessage(agent codersdk.WorkspaceAgent, url string) string {
 		m += " and " + agent.TroubleshootingURL
 	}
 	return m
+}
+
+// safeDuration returns a-b. If a or b is nil, it returns 0.
+// This is because we often dereference a time pointer, which can
+// cause a panic. These dereferences are used to calculate durations,
+// which are not critical, and therefor should not break things
+// when it fails.
+// A panic has been observed in a test.
+func safeDuration(a, b *time.Time) time.Duration {
+	if a == nil || b == nil {
+		return 0
+	}
+	return a.Sub(*b)
 }
 
 type closeFunc func() error

